### ISSUE TITLE
`reduce` for Python 3.x

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4066]
|<img alt="cvaliente's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| cvaliente reported|
|-|-|
|Trac ID|trac#4066|
|Type|enhancement|
|Created|2009-10-29 09:40:16Z|
|Branch|https://github.com/twisted/twisted/tree/reduce-4066-2|

This patch eliminates the DeprecationWarning for "reduce() not supported in 3.x; use functools.reduce()"

Attachments:

* [reduce.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/352/352e04be1f73fd02137de7409de203acf127a3a7/5337afcb4aef2abb5f15c80e153a4034691a0e70.diff) (4514 bytes) - added by cvaliente on 2009-10-29 09:40:30Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__4066 4066
type__enhancement enhancement
reporter__cvaliente cvaliente
priority__normal normal
milestone__ 
branch__branches_reduce_4066_2 branches/reduce-4066-2
branch_author__exarkun__cvaliente exarkun, cvaliente
status__closed closed
resolution__fixed fixed
component__core core
keywords__py3k py3k
time__1256809216000000 1256809216000000
changetime__1259889024000000 1259889024000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

(In [27486]) Branching to 'reduce-4066'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

Seems generally to be the right idea.  I think we can dump `reduce` into `twisted.python.compat` though, and remove some of the uses of reduce.  I put the patch in the branch and made those changes, plus updated some copyright dates.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

(In [27492]) Branching to 'reduce-4066-2'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

Okay, I did like I said, someone else should take a look at this now.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_7"></a>|@radix set owner to @exarkun|
|-|-|

Looks good, please merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun set status to `closed`|
|-|-|

(In [27667]) Merge reduce-4066-2

Author: exarkun, cvaliente
Reviewer: radix
Fixes: [#4066](https://github.com/twisted/twisted/issues/4066)

Try to use `functools.reduce` when it is available, instead of the builtin
`reduce`, which is removed in Python 3.x.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
